File: test/lib-test-cases/tools/uid testcase generator/murmur_hash3.py
"""Python port of MurmurHash3 x86_32 by Austin Appleby (public domain).

Origin: https://github.com/aappleby/smhasher
"""

import struct


def _rotl32(x: int, r: int) -> int:
    return ((x << r) | (x >> (32 - r))) & 0xFFFFFFFF


def _fmix32(h: int) -> int:
    h = (h ^ (h >> 16)) & 0xFFFFFFFF
    h = (h * 0x85ebca6b) & 0xFFFFFFFF
    h = (h ^ (h >> 13)) & 0xFFFFFFFF
    h = (h * 0xc2b2ae35) & 0xFFFFFFFF
    h = (h ^ (h >> 16)) & 0xFFFFFFFF
    return h


def murmur_hash3_x86_32(key: bytes, seed: int = 0) -> bytes:
    """Return the MurmurHash3 x86_32 hash of *key* as 4 little-endian bytes."""
    length = len(key)
    n_blocks = length // 4

    h1 = seed & 0xFFFFFFFF
    c1 = 0xcc9e2d51
    c2 = 0x1b873593

    # Body: process 4-byte blocks
    for i in range(n_blocks):
        k1 = struct.unpack_from('<I', key, i * 4)[0]
        k1 = (k1 * c1) & 0xFFFFFFFF
        k1 = _rotl32(k1, 15)
        k1 = (k1 * c2) & 0xFFFFFFFF
        h1 ^= k1
        h1 = _rotl32(h1, 13)
        h1 = (h1 * 5 + 0xe6546b64) & 0xFFFFFFFF

    # Tail: remaining 1-3 bytes (mirrors the C++ switch fallthrough)
    tail = n_blocks * 4
    k1 = 0
    remainder = length & 3

    if remainder >= 3:
        k1 ^= key[tail + 2] << 16
    if remainder >= 2:
        k1 ^= key[tail + 1] << 8
    if remainder >= 1:
        k1 ^= key[tail]
        k1 = (k1 * c1) & 0xFFFFFFFF
        k1 = _rotl32(k1, 15)
        k1 = (k1 * c2) & 0xFFFFFFFF
        h1 ^= k1

    # Finalization
    h1 ^= length
    h1 = _fmix32(h1)

    return struct.pack('<I', h1)


if __name__ == '__main__':
    # Quick sanity checks against known MurmurHash3 reference values
    assert murmur_hash3_x86_32(b'', seed=0) == struct.pack('<I', 0x00000000)
    assert murmur_hash3_x86_32(b'', seed=1) == struct.pack('<I', 0x514E28B7)
    assert murmur_hash3_x86_32(b'', seed=0xFFFFFFFF) == struct.pack('<I', 0x81F16F39)
    assert murmur_hash3_x86_32(b'\xff\xff\xff\xff', seed=0) == struct.pack('<I', 0x76293B50)
    assert murmur_hash3_x86_32(b'\x21\x43\x65\x87', seed=0) == struct.pack('<I', 0xF55B516B)
    assert murmur_hash3_x86_32(b'\x21\x43\x65\x87', seed=0x5082EDEE) == struct.pack('<I', 0x2362F9DE)
    assert murmur_hash3_x86_32(b'\x21\x43\x65', seed=0) == struct.pack('<I', 0x7E4A8634)
    assert murmur_hash3_x86_32(b'\x21\x43', seed=0) == struct.pack('<I', 0xA0F7B07A)
    assert murmur_hash3_x86_32(b'\x21', seed=0) == struct.pack('<I', 0x72661CF4)
    print('All self-tests passed.')

File: firmware_updater/updater/src/org/selfbus/updater/MurmurHash3.java
package org.selfbus.updater;

// MurmurHash3 was written by Austin Appleby, and is placed in the public
// domain. The author hereby disclaims copyright to this source code.

// origin https://github.com/aappleby/smhasher


// Java port of the x86_32 variant.
public class MurmurHash3 {

    private static int rotl32(final int x, final int r) {
        return (x << r) | (x >>> (32 - r));
    }

    private static int getblock32(final byte[] data, final int offset) {
        return (data[offset]     & 0xFF)
             | (data[offset + 1] & 0xFF) << 8
             | (data[offset + 2] & 0xFF) << 16
             | (data[offset + 3] & 0xFF) << 24;
    }

    private static int fmix32(int h) {
        h ^= h >>> 16;
        h *= 0x85ebca6b;
        h ^= h >>> 13;
        h *= 0xc2b2ae35;
        h ^= h >>> 16;
        return h;
    }

    /**
     * Computes the MurmurHash3 x86_32 hash of the given byte array.
     *
     * @param key  input data
     * @param seed seed value
     * @return 32-bit hash as a long (unsigned interpretation) or use as int
     */
    @SuppressWarnings("fallthrough")
    public static int murmurHash3_x86_32(final byte[] key, final int seed) {
        final int len = key.length;
        final int nBlocks = len / 4;

        int h1 = seed;

        final int c1 = 0xcc9e2d51;
        final int c2 = 0x1b873593;

        // body
        for (int i = 0; i < nBlocks; i++) {
            int k1 = getblock32(key, i * 4);

            k1 *= c1;
            k1 = rotl32(k1, 15);
            k1 *= c2;

            h1 ^= k1;
            h1 = rotl32(h1, 13);
            h1 = h1 * 5 + 0xe6546b64;
        }

        // tail
        int k1 = 0;
        final int tailOffset = nBlocks * 4;

        switch (len & 3) {
            case 3: k1 ^= (key[tailOffset + 2] & 0xFF) << 16;
                    // fall through
            case 2: k1 ^= (key[tailOffset + 1] & 0xFF) << 8;
                    // fall through
            case 1: k1 ^= (key[tailOffset] & 0xFF);
                    k1 *= c1;
                    k1 = rotl32(k1, 15);
                    k1 *= c2;
                    h1 ^= k1;
                    break;
            default:
                break;
        }

        // finalization
        h1 ^= len;
        h1 = fmix32(h1);

        return h1;
    }
}
